Use networkx

jo-mueller opened the pull request:
Hi Robert @haesleinhuepf ,

this pull request replaces drawing the data flow graph with igraph with a networkx-based solution so you can switch the license to BSD3 :)

Major changes:
- I replaced the matplotlibwidget with my extended version, so you can use the toolbar as well to pan/zoom/etc the graph in case it becomes too large
- I added a function that traverses the currently installed workflow and creates a `nx.DiGrap` (directed graph) from it. I made the graph and the graphwidget members of `WorkflowWidget` to facilitate handling and detect updates/changes to the workflow. Thus, the graph is really only redrawn when a change to the graph has been made. *Note*: Currently, this is only checked through the length of the workflow - If an element of the workflow were to be replaced, this wouldn't be detected.
- I replaced the license with a BSD-3 clause since networkx is also BSD-3 licensed.

Here's how it looks:
![new_widget](https://user-images.githubusercontent.com/38459088/152678604-0367b073-25e8-4585-9129-3be77c81e308.png)

Side notes:
-  It *seems* as if the widgets created from the assistant have some weird Size-Policy as they collapse the side-bar to a fixed size that cannot be changed anymore. If I hide the assistant-created widget, the horizontal size of the sidebar can be changed. Is this a bug or a feature?
- I would like to take this a little further and install callbacks on the graph canvas which connect to the workflow. Thus blicking on the points in the graph could trigger highlighting the layers or associated widgets for a particular operation , displaying the function parameters, etc.

Let me know what you think :)

jo-mueller committed:
created function to create networkx grapph from workflow

jo-mueller committed:
implemented drawing of networkx graph

- made graphwidget member of Workflow-widget
- muted igraph-based drawing

jo-mueller committed:
Replaced matplotlibwidget

- new widget allows mpl tool bar (zooming, panning, etc)
- made networkx member of widget so a change in the number of nodes can be detected upon updating and the plot is not redrawed more often than really necessary

jo-mueller committed:
changed license to BSD-3

jo-mueller committed:
added networkx to the dependencies

jo-mueller committed:
update graph when method has changed

jo-mueller commented:
> Note: Currently, this is only checked through the length of the workflow - If an element of the workflow were to be replaced, this wouldn't be detected.

Fixed in https://github.com/haesleinhuepf/napari-workflow-inspector/pull/3/commits/3cdb7d080239c7db12377d0340f24d2d8742e2fb. You can now change methods and the graph is updated accordingly.

jo-mueller committed:
removed unnecessary function and improved plot layout

jo-mueller commented:
Made a few improvements to the plot layout in https://github.com/haesleinhuepf/napari-workflow-inspector/pull/3/commits/5e8d71e806bc3c17c56cc195748ebaf1c3e0a7a2.

haesleinhuepf commented on src/napari_workflow_inspector/_dock_widget.py:
```diff
@@ -1,26 +1,57 @@
-        self.axes = self.fig.add_subplot(111)
-
-        super(MplCanvas, self).__init__(self.fig)
+        self.axes.set_facecolor('#262930')
```
I'm speculating that this line should be removed to have a transparent background,

haesleinhuepf replied:
I'm speculating that this line should be removed to have a transparent background,

haesleinhuepf commented on src/napari_workflow_inspector/_dock_widget.py:
```diff
@@ -118,48 +154,13 @@ def html(text):
-                                vertex_size = 10,
-                                vertex_label_dist = 10,
-                                vertex_label_size = 10,
-                                edge_color = "#dddddd",
```
I'm speculating that your code could profit from having edges in light-grey

haesleinhuepf commented on src/napari_workflow_inspector/_dock_widget.py:
```diff
@@ -118,48 +154,13 @@ def html(text):
-                        self.graph_layout = idfg.layout_auto()
-
-                    # workaround as vertex_label_color above doesn't work :-(
-                    matplotlib.rcParams['text.color'] = '#dddddd'
```
I _think_ also this line should be kept somewhere to make the text readable on dark background.

haesleinhuepf reviewed (changes requested):
Hi Johannes @jo-mueller ,

this is so awesome! Thanks for your initiative. Unfortunately, some of the former functionality is lost through this update. I left some comments in the code where I think you could take stuff over from the former implementation. 

* The white background is not optimal in a black napari window.
* the old graph showed the status of the operations/layers in color, red: not valid, green: valid https://github.com/haesleinhuepf/napari-workflow-inspector/blob/48b2459bbfdd46ae665c74b81007617cc1c4be44/src/napari_workflow_inspector/_dock_widget.py#L89-L94

Furthermore, there is currently a bug in the visualization (or in the workflows?): When I setup a workflow on blobs.tif using the assistant: Gaussian Blur > Threshold Otsu > Voronoi-Otsu-Labeling and then change the sigma of the Gaussian blur, the graph is changing its architecture... Not sure why.

Would you mind iterating over the code once more and testing it a bit more in detail?

Again, big thanks for working on this!

Best,
Robert

jo-mueller commented:
> the graph is changing its architecture... Not sure why.

This I think I can answer: It's not a bug, it's a feature. Networkx provides a few layouts to draw the graphs, but the optimal layout depends very much on the size/shape of the graph, so this can (and I think should) change to a certain extent as the graph gets more and more complex.

I'll see how the other layouts look like.

As for the rest, thanks for the feedback, I'll have a look at it 👍

jo-mueller committed:
put node-position determination in separate command

jo-mueller committed:
introduced class for clickable workflow steps

This eases handling colors of single workflow steps, i.e. to indicate the validity of associated layers.

jo-mueller committed:
Integrated layer validation scheme

jo-mueller committed:
made graph display prettier and implemented on-pick callback to Workflow

jo-mueller commented:
Hi @haesleinhuepf ,

would you have a look at it again? I fixed your remarks from above and added another cool feature. Some comments ahead:

> the graph is changing its architecture... Not sure why.

After reading a bit more on this: The layout determination algorithms for graphs in networkx are not deterministic - hence, the graph will look different every time it is created. Some of the layouts provide options to keep some nodes at their original position, but I'm not sure whether this is actually desirable. It could see this becoming messy if the graph is not allowed to adapt when, for instance, two images are combined. It would also be possible to add a little  dropdown to change the layout, if desired.

However, it is kind of strange that the layout changes when parameters are adjusted - this is not the case in my version. Maybe it is fixed after the latest changes!

> The white background is not optimal in a black napari window.

Fixed!

> the old graph showed the status of the operations/layers in color, red: not valid, green: valid

I implemented this as well. I change the color scheme to cyan (valid) and orange (invalid) to account for colorblindness. Just as a sidenode: I think we now both implemented some sort of graph traversal to update the workflow in the different views (from source and visual). I thhink it could make sense to unify this lateron.

Last but not least: Try clicking on the nodes in the workflow graph :)

jo-mueller committed:
Fixed bugs

-> Edge color stays the same
-> Removed unnecessary line in node picking

jo-mueller commented:
Hi Robert @haesleinhuepf ,

I fixed the bug with the changing line color. Let me know what you think.

haesleinhuepf commented:
Awesome, thanks Johannes @jo-mueller , I'm going to do the last changes we discussed myself (changing colours to green-magenta).

Also, next time, you submit a PR to any of "my" napari plugins, consider to add yourself to the [list of authors](https://github.com/haesleinhuepf/napari-workflow-inspector/blob/057775b1f2044de9cb3756a429cebed27836325b/setup.cfg#L4). You are very welcome there.

Thanks again! Fantastic work :-)

haesleinhuepf commented:
Oh and one more hint: Next time, it would be great if you could also update the documentation in the readme. Thanks!

jo-mueller commented:
Noted - thanks for merging!
